meetings/working-groups/collection-literals/CL-2023-04-05.md

```diff
@@ -0,0 +1,94 @@
+# Collection literals working group meeting for April 4, 2023
+
+## Agenda
+
+* Allow efficient construction of collections that are typically represented as arrays, particularly `ImmutableArray<T>`.
+
+## Discussion
+
+We discussed possible _construction method_ patterns that would allow the compiler to write directly to the underlying storage for collections such as `ImmutableArray<T>` and `List<T>`, and avoid extra allocations or copying in cases where the collection has a _known length_.
+
+There were two promising approaches that provide equivalent capability, with the difference of whether the underlying storage is allocated by the caller or by the collection.
+
+### Pass ownership of array
+One possible approach is a _construction method_ that takes an array as an argument and uses that array directly as the underlying storage of the collection, with the caller passing ownership of the array to the collection instance.
+
+For instance, given the following method for creating an `ImmutableArray<T>`:
+```csharp
+// Create ImmutableArray<T> using the array as the underlying storage.
+public static ImmutableArray<T> AsImmutableArray<T>(T[] array)
+{
+    // ...
+}
+```
+
+The compiler could translate `ImmutableArray<T> result = [x, ..e, y];` to:
+```csharp
+T[] __array = new T[e.Count + 2];
+
+int __i = 0;
+__array[__i++] = x;
+foreach (T __t in e)
+    __array[__i++] = __t;
+__array[__i++] = y;
+
+ImmutableArray<T> result = AsImmutableArray(__array);
+```
+
+### Write directly to storage span
+An alternative is a _construction method_ that creates a collection of a given size and returns the collection _and_ a span that refers to the underlying storage, and the caller populates the collection using the span.
+
+For instance, given the following method for creating a `List<T>`:
+```csharp
+// Create List<T> and return a span to the underlying storage.
+public static List<T> CreateList<T>(int size, out Span<T> span)
+{
+    // ...
+}
+```
+
+The compiler could translate `List<T> result = [x, ..e, y];` to:
+```csharp
+Span<T> __span;
+List<T> result = CreateList(e.Count + 2, out __span);
+
+int __i = 0;
+__span[__i++] = x;
+foreach (T __t in e)
+    __span[__i++] = __t;
+__span[__i++] = y;
+```
+
+### Runtime libraries
+
+The BCL could provide _construction methods_ for `ImmutableArray<T>` and `List<T>`, and perhaps a few additional collection types where the underlying storage can be represented as a span.
```

Comment:
Should we mention that the BCL may do so already, and it sounded likely from how @stephentoub described it?

`ImmutableArray.CreateUnsafe<T>` is waiting for API review (https://github.com/dotnet/runtime/issues/83141)

I didn't find an issue for a single method for the "set count and request span" construction method that he mentioned. That would have to be a new method combining `CollectionsMarshal.SetCount<T>(List<T>, int)` approved at https://github.com/dotnet/runtime/issues/55217 and the shipped [`CollectionsMarshal.AsSpan<T>(List<T>)`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan), and this use case is listed in the SetCount issue.

Reply:
I'll leave as is, since it looks like those are API changes are only proposed currently.